crates/circuit/src/circuit_data.rs

```diff
@@ -171,8 +170,8 @@ impl CircuitData {
             }
         }
         if let Some(data) = data {
-            self_.reserve(py, reserve);
-            self_.extend(py, data)?;
+            self_.reserve(reserve);
+            Python::with_gil(|py| -> PyResult<()> { self_.extend(py, data) })?;
```

Comment:
It would be nice if we could also do this without the py token to enable copying circuits. What do you think about moving the `with_gil` inside `extend` in a way that we check if an instruction has parameter expressions and only then acquire the GIL? Maybe this has too much overhead and we can just wait for the Rust parameter expressions, too 😄

Reply:
I left it here because `extend()` was being passed a `Bound<PyAny>` I can  drop the py here and just use it from the bound.